pkg/alertmanager/distributor/distributor.go

```diff
@@ -0,0 +1,317 @@
+	"github.com/opentracing/opentracing-go"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/weaveworks/common/httpgrpc"
+	"github.com/weaveworks/common/httpgrpc/server"
+	"github.com/weaveworks/common/user"
+
+	"github.com/cortexproject/cortex/pkg/alertmanager/alertmanagerpb"
+	"github.com/cortexproject/cortex/pkg/ring"
+	"github.com/cortexproject/cortex/pkg/ring/client"
+	"github.com/cortexproject/cortex/pkg/tenant"
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/cortexproject/cortex/pkg/util/services"
+)
+
+// RingOp is the operation used for reading/writing to the alertmanagers.
+var RingOp = ring.NewOp([]ring.IngesterState{ring.ACTIVE}, func(s ring.IngesterState) bool {
+	// Only ACTIVE Alertmanager get requests. If instance is not ACTIVE, we need to find another Alertmanager.
+	return s != ring.ACTIVE
+})
+
+// Config contains the configuration required to
+// create a Distributor
+type Config struct {
+	RemoteTimeout      time.Duration            `yaml:"remote_timeout"`
+	AlertmanagerClient AlertmanagerClientConfig `yaml:"alertmanager_client"`
+}
+
+// RegisterFlags adds the flags required to config this to the given FlagSet
+func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
+	cfg.AlertmanagerClient.RegisterFlagsWithPrefix("alertmanager-distributor.alertmanager-client", f)
+	f.DurationVar(&cfg.RemoteTimeout, "alertmanager-distributor.remote-timeout", 2*time.Second, "Timeout for downstream alertmanagers.")
+}
+
+// Distributor forwards requests to individual alertmanagers.
+type Distributor struct {
+	services.Service
+
+	cfg              Config
+	requestsInFlight sync.WaitGroup
+
+	alertmanagerRing        ring.ReadRing
+	alertmanagerClientsPool AlertmanagerClientsPool
+
+	// Manager for subservices (Alertmanager Ring)
+	subservices        *services.Manager
+	subservicesWatcher *services.FailureWatcher
+
+	logger log.Logger
+
+	receivedRequests  *prometheus.CounterVec
+	amSends           *prometheus.CounterVec
+	amSendFailures    *prometheus.CounterVec
+	replicationFactor prometheus.Gauge
+}
+
+// New constructs a new Distributor
+func New(cfg Config, alertmanagersRing *ring.Ring, alertmanagerClientsPool AlertmanagerClientsPool, logger log.Logger, reg prometheus.Registerer) (d *Distributor, err error) {
+	if logger == nil {
+		logger = log.NewNopLogger()
+	}
+
+	if alertmanagerClientsPool == nil {
+		alertmanagerClientsPool = newAlertmanagerClientsPool(client.NewRingServiceDiscovery(alertmanagersRing), cfg.AlertmanagerClient, logger, reg)
+	}
+
+	d = &Distributor{
+		cfg:                     cfg,
+		logger:                  logger,
+		alertmanagerRing:        alertmanagersRing,
+		alertmanagerClientsPool: alertmanagerClientsPool,
+		//replication:             alertmanagersRing.ReplicationFactor(),
+	}
+
+	d.initMetrics(reg)
+	d.replicationFactor.Set(float64(alertmanagersRing.ReplicationFactor()))
```

Comment:
Do we need this metric at all? We currently have it only for the distributor (the series distributor) and we don't have it for the store-gateway, and I don't think we felt missing it.